R/Data-validity.R

```diff
@@ -1,35 +1,120 @@
-#' Internal Helper Functions for Validation of `GeneralData` Objects.
+#' Internal Helper Functions for Validation of `GeneralData` and `Data` Objects.
 #'
 #' @description `r lifecycle::badge("stable")`
-#' 
+#'
 #' These functions are only used internally to validate the format of an input
-#' [`GeneralData-class`] object and therefore not exported.
+#' [`GeneralData-class`] and [`Data-class`] object and therefore not exported.
 #'
 #' @name validate_data
 #' @param object (`GeneralData`)\cr object to validate.
 #' @return A character vector with the validation failure messages,
 #' or `TRUE` in case validation passes.
 NULL
 
+#' @describeIn validate_data helper function.
+#' It returns `TRUE` if `dose` is unique per `cohort`.
+#' @param dose (`numeric`)\cr vector of doses.
+#' @param cohort (`integer`)\cr vector of cohorts corresponding to `doses`.
+#'
+h_is_dose_unique <- function(dose, cohort) {
+  all(tapply(dose, cohort, function(d) length(unique(d)) == 1L))
+}
+
 #' @describeIn validate_data validates that the [`GeneralData-class`] object contains
 #'   unique `ID`, non-negative `cohort` indices and
 #'   `ID` and `cohort` vectors are of the same length equal to `nObs`.
 #'
 validate_subjects <- function(object) {
   o <- Validate()
-  
+
   o$check(test_int(object@nObs), "nObs must be of type integer of length 1")
-  
-  o$check(is.integer(object@ID), "ID must be of type integer")
+
+  o$check(
+    test_integer(object@ID, len = object@nObs),
+    "ID must be of type integer and length nObs"
+  )
   o$check(all(!duplicated(object@ID)), "IDs must be unique")
-  
-  o$check(is.integer(object@cohort), "cohort must be of type integer")
+
+  o$check(
+    test_integer(object@cohort, len = object@nObs),
+    "cohort must be of type integer and length nObs"
+  )
   o$check(all(object@cohort >= 0), "cohort indices must be non-negative")
   o$check(!is.unsorted(object@cohort, strictly = FALSE), "cohort indices must be sorted")
-  
+
+  o$result()
+}
+
+#' @describeIn validate_data validates that the [`Data-class`] object contains
+#' valid elements with respect to their types, dependency and length.
+#'
+validate_data <- function(object) {
```

Comment:
this function is too long (> 50 lines as rule of thumb) - let's think about logical parts we could break this into